feat: add toBeVisible matcher

fluiddot opened the pull request:
<!--

Thanks for your interest in the project. Bugs filed and PRs submitted are appreciated!

Please make sure that you are familiar with and follow the Code of Conduct for this project (found in the CODE_OF_CONDUCT.md file).

Please fill out the information below to expedite the review and (hopefully) merge of your pull request!

-->

**What**:

<!-- What changes are being made? (What feature/bug is being fixed here?) -->

Add `toBeVisible` matcher following a similar logic as the `jest-dom` matcher ([reference](https://github.com/testing-library/jest-dom#tobevisible)).

**Why**:

<!-- Why are these changes necessary? -->

This matcher provides a useful way to verify that a queried element is actually visible.

**How**:

<!-- How were these changes implemented? -->

The matcher has been introduced based on the implementation of other matchers and trying to reproduce a similar logic as the `jest-dom` matcher ([reference](https://github.com/testing-library/jest-dom/blob/main/src/to-be-visible.js)), but taking into account the React Native nuances.

**Checklist**:

<!-- Have you done all of these things?  -->
<!-- Add "(N/A)" to the end of each line that's irrelevant to your changes -->
<!-- to check an item, place an "x" in the box like so: "- [x] Documentation" -->

- [x] Documentation added to the
      [docs](https://github.com/testing-library/jest-native/README.md)
- [x] Typescript definitions updated
- [x] Tests
- [x] Ready to be merged <!-- In your opinion -->

<!-- feel free to add additional comments -->

dcalhoun commented on README.md:
````diff
@@ -298,6 +299,78 @@ expect(queryByText('Hello World')).toHaveStyle([{ color: 'black' }, { fontWeight
 expect(queryByText('Hello World')).not.toHaveStyle({ color: 'white' });
 ```
 
+### `toBeVisisble`
````
```suggestion
### `toBeVisible`
```
It would appear each occurrence of "visible" is misspelled in this file. It would create a lot of unnecessary noise for me to call out each one, but we should likely fix the spelling of each misspelling for this file.

fluiddot replied:
Oh, my bad, I spelled it wrong and copy/paste it all over the file 😅 , I'll fix it as soon as possible, thanks for spotting it 🙇 .

fluiddot replied:
I fixed this in https://github.com/testing-library/jest-native/commit/96c6ac525e1e547540fbeb9b27bc8056d214cb95.

dcalhoun commented on src/__tests__/to-be-visible.js:
```diff
@@ -0,0 +1,45 @@
+  });
+
+  test.each([
+    ['View with 0 opacity', <View testID="test" style={{ opacity: 0 }} />],
```
```js
test.each([
    // ...
    ['View with display none', <View testID="test"  style={{ display: 'none' }} />],
    // ...
]);
```

To better mirror [`jest-dom`'s `toBeVisible`](https://github.com/testing-library/jest-dom#tobevisible) and cover additional contexts where an element would not be visible, WDYT about checking for `display: 'none'` as well?

fluiddot replied:
Good point, I overlooked the `display` prop for this case so I'll include it into the condition for checking the visibility.

fluiddot replied:
I added the suggestion in https://github.com/testing-library/jest-native/commit/e7afe068719863c95873c911a63e3f52c9004b46.

dcalhoun commented on src/to-be-visible.js:
```diff
@@ -0,0 +1,40 @@
+function isElementVisible(element) {
+  return (
+    isStyleVisible(element) &&
+    (!element.parent || isStyleVisible(element.parent)) &&
```
I may be misinterpreting the code, but this appears to only check one ancestor. I.e. if the subject element is a grandchild, this will not verify the grandparent's visibility. Could we leverage recursion and achieve's [`jest-dom`'s](https://github.com/testing-library/jest-dom#tobevisible) "its parent element is also visible (and so on up to the top of the DOM tree)" ability? 

```js
function isElementVisible(element) {
  return (
    isStyleVisible(element) &&
    isAttributeVisible(element) &&
    (!element.parent || isElementVisible(element.parent))
  );
}
```

fluiddot replied:
Right, the idea is to traverse all element's ancestors in order to check that both style and attribute are visible. In fact, the recursion to achieve this is already covered by the following line:

https://github.com/fluiddot/jest-native/blob/feat/to-be-visible/src/to-be-visible.js#L21

I think I incorrectly added this line as it's not really necessary, I'll remove it as soon as possible, thanks for catching this 🙇 .

fluiddot replied:
I fixed this issue in https://github.com/testing-library/jest-native/commit/96c6ac525e1e547540fbeb9b27bc8056d214cb95.

fluiddot replied:
I fixed this issue in https://github.com/testing-library/jest-native/commit/5dd7ffe7abebf82a5fe810c24da02baf1973aa21.

dcalhoun reviewed (commented):
Great work! This matcher would be quite useful. I left a few comments to consider.

fluiddot commented:
> Great work! This matcher would be quite useful. I left a few comments to consider.

Thank you very much @dcalhoun for taking a look at this PR 🙇.

Following your comments and suggestions, I've applied some updates to the changes. At this point, the PR can be considered as ready to be reviewed in order to include the new `toBeVisible` matcher.

fluiddot commented:
Looks like most of the unit tests are failing with the following error:
```
TypeError: _interopRequireDefault is not a function
```

However, I managed to reproduce the same issue by running the tests in the `main` branch, so not sure whether these changes might be the cause 🤔 .

dcalhoun commented:
@mdjastrzebski @thymikee would either of you be willing to review this work please? Totally understand if you do not have time currently. Just wanted to check in. Thanks! 🙇🏻

GuglielmoFelici commented:
Hi all! Any update on this? Would be quite useful 🙌

mdjastrzebski commented:
I will try to work on that next week

Norfeldt commented:
> I will try to work on that next week

Don't know if this would be of any help @mdjastrzebski . I'm using this:

```
expect.extend({
  // should be replacable once this PR gets merged https://github.com/testing-library/jest-native/pull/73
  toHaveZeroOpacity(element) {
    let pass = element.props.style?.opacity === 0
    let parent = element.parent
    while (parent && pass === false) {
      if (parent.props.style?.opacity === 0) pass = true
      parent = parent?.parent
    }

    return {
      message: () => 'expected element to be hidden',
      pass,
    }
  },
})

```


```
import { render } from '@testing-library/react-native'
import * as React from 'react'
import { View, Text } from 'react-native'

describe('extended `expect` methods', () => {
  describe('.toHaveZeroOpacity', () => {
    it('should pass when element is has zero opacity', () => {
      const utils = render(<View style={{ opacity: 0 }} testID="TestId__VIEW" />)
      expect(utils.getByTestId('TestId__VIEW')).toHaveZeroOpacity()
    })

    it('should NOT pass when element is has non-zero opacity', () => {
      const utils = render(<View style={{ opacity: 0.001 }} testID="TestId__VIEW" />)
      expect(utils.getByTestId('TestId__VIEW')).not.toHaveZeroOpacity()
    })

    it('should pass when some parent element is has zero opacity', () => {
      const utils = render(
        <View style={{ opacity: 0 }}>
          <View>
            <Text>invisible</Text>
          </View>
        </View>
      )
      expect(utils.getByText(/invisible/i)).toHaveZeroOpacity()
    })

    it('should pass when some parent element is has zero opacity', () => {
      const utils = render(
        <View>
          <View>
            <Text>invisible</Text>
          </View>
        </View>
      )
      expect(utils.getByText(/invisible/i)).not.toHaveZeroOpacity()
    })
  })
})
```

mdjastrzebski commented on README.md:
````diff
@@ -298,6 +299,107 @@ expect(queryByText('Hello World')).toHaveStyle([{ color: 'black' }, { fontWeight
+toBeVisible();
+```
+
+Check that the given element is visible.
````
```suggestion
Check that the given element is visible to the user.
```

mdjastrzebski commented on README.md:
```diff
@@ -298,6 +299,107 @@ expect(queryByText('Hello World')).toHaveStyle([{ color: 'black' }, { fontWeight
+
+Check that the given element is visible.
+
+An element is visible if **all** the following conditions are met:
```
It seems that we have a related task in RNTL for adding option to ignore hidden elements in queries. It seems that these two PRs should share the same logic for visibility check.
Pls check the other PR: https://github.com/callstack/react-native-testing-library/pull/1064/files#diff-bc7c1ed436746d189e6b39ab5fdd4e2092589eb23da9a67e68e067a884231d73

mdjastrzebski replied:
This is no longer relevant, as I have learned implementing the [`isInaccessible`](https://github.com/callstack/react-native-testing-library/blob/cc52c610fc9b9f8686f8e6204454b3f6a13888b3/src/helpers/accessiblity.ts#L5) function for RNTL, that these two are similar but not the same, as e.g. `opacity: 0` does not exclude from accessibility tree, but would make the element invisible.

mdjastrzebski commented on src/to-be-visible.js:
```diff
@@ -0,0 +1,39 @@
+}
+
+function isAttributeVisible(element) {
+  return element.type !== 'Modal' || element.props.visible !== false;
```
Directly checking modal type looks kinda hacky, maybe we could check some accessiblity props here: `accessibilityElementsHidden`, etc

moxspoy replied:
does it support bottom sheet from [gorhom](https://github.com/gorhom/react-native-bottom-sheet) ?

mdjastrzebski replied:
Hard to say without checking, depends on how well that library supports accessibility props. I think we should first concentrate on supporting built-in components, e.g. built-in `Modal` from RN, and handle the external libraries afterwards on a case-by-case basis, as in some cases the proper place to support them will be here, but in other case it might be by submitting the PR to the library authors to improve their accessibility support.

mdjastrzebski replied:
@fluiddot I wonder how `Modal` actually performs hidding of its children when `visibility=false`. Instead of relying on querying the element for type and `visible` prop, we could check the underlying code. Maybe it just does not render the children at all.

fluiddot replied:
> Instead of relying on querying the element for type and visible prop, we could check the underlying code. Maybe it just does not render the children at all.

@mdjastrzebski I checked when testing that `Modal` component doesn't actually render children, so this matcher will throw an exception when using it for children elements of a not visible modal.

fluiddot replied:
> Directly checking modal type looks kinda hacky, maybe we could check some accessiblity props here: accessibilityElementsHidden, etc

@mdjastrzebski good point. To be honest, I added the modal type check as a special case, since it's likely that we want to check the visibility of a modal using this matcher. In any case, I'm open to reconsider this condition in favor of a better approach or even not including it.

Regarding the accessibility props, I went ahead and updated the matcher to also check them in [this commit](https://github.com/testing-library/jest-native/pull/73/commits/33482121b8b15ff84aaf6a104c3197d18f70aa59), similar to the logic of [`isInaccessible` function](https://callstack.github.io/react-native-testing-library/docs/api#isinaccessible).

mdjastrzebski reviewed (changes requested):
@fluiddot I've added some comments regarding sharing logic with a related PR in RNTL. We also removed `ramda` in the mean time so pls update your imports.

tomwaitforitmy commented:
I would be very happy to use this! 👍

fluiddot commented:
> @fluiddot Do you have capacity to work on updating this PR?

Hey @mdjastrzebski, my apologies for not being very active in the PR. I'd like to resume the work on this and review the feedback as soon as possible, but my bandwidth nowadays is usually low. If someone is willing to take over this, please feel to do it. Otherwise, I'll try to make some progress in the next weeks.

mdjastrzebski commented on src/__tests__/to-be-visible.js:
```diff
@@ -0,0 +1,60 @@
+import { render } from '@testing-library/react-native';
+
+describe('.toBeVisible', () => {
+  test.each([
```
Github/link seems to be complaining about missing `key` prop, Could you refactor these tests to be regular. not `.each` tests for readability.

fluiddot replied:
I fixed this issue when migrating the test file to TypeScript in [this commit](https://github.com/testing-library/jest-native/pull/73/commits/861cc729f7857aed8d05f120e10aa293b4829da0).

mdjastrzebski commented on src/to-be-visible.js:
```diff
@@ -0,0 +1,39 @@
+
+function isStyleVisible(element) {
+  const style = element.props.style || {};
+  const { display = 'flex', opacity = 1 } = Array.isArray(style) ? mergeAll(style) : style;
```
Use `StyleSheet.flatten` to calculate flat `style` object.

fluiddot replied:
Good catch, I've applied this suggestion in [this commit](https://github.com/testing-library/jest-native/pull/73/commits/2335a11668f03b87d8cc1287096b6dcfbf378eb3).

mdjastrzebski commented on src/to-be-visible.js:
```diff
@@ -0,0 +1,39 @@
+
+function isStyleVisible(element) {
+  const style = element.props.style || {};
+  const { display = 'flex', opacity = 1 } = Array.isArray(style) ? mergeAll(style) : style;
```
No need for default values, if we later compare them to known values.
```suggestion
  const { display, opacity } = Array.isArray(style) ? mergeAll(style) : style;
```

fluiddot replied:
Agreed, I've applied this suggestion in [this commit](https://github.com/testing-library/jest-native/pull/73/commits/2335a11668f03b87d8cc1287096b6dcfbf378eb3).

mdjastrzebski commented:
@fluiddot Overall the current implementation looks solid. I think it's sensible approach to base it on [Jest DOM matcher code](https://github.com/testing-library/jest-dom/blob/main/src/__tests__/to-be-visible.js) with some tweaks to adjust for different props in RN than in Web.

fluiddot commented:
> @fluiddot would you be able to rebase the code and migrate it to typescript as the current codebase uses it now.

@mdjastrzebski 😅 Sorry for the long delay on addressing the comments of the PR and migrating the changes to TypeScript. After the recent changes I pushed, the PR should be ready for another review. Let me know if there's anything else I can contribute with in order to conclude the feature. Thank you very much for the help 🙇 .

mdjastrzebski commented on src/__tests__/to-be-visible.tsx:
```diff
@@ -0,0 +1,127 @@
+    expect(getByTestId('test')).toBeVisible();
+  });
+
+  test('handles view within innacessible view (Android)', () => {
```
```suggestion
  test('handles view within inaccessible view (Android)', () => {
```

mdjastrzebski commented on src/__tests__/to-be-visible.tsx:
```diff
@@ -0,0 +1,127 @@
+    expect(getByTestId('test')).not.toBeVisible();
+  });
+
+  test('handles innacessible view (Android)', () => {
```
```suggestion
  test('handles inaccessible view (Android)', () => {
```

mdjastrzebski commented on src/__tests__/to-be-visible.tsx:
```diff
@@ -0,0 +1,127 @@
+    expect(getByTestId('test')).toBeVisible();
+  });
+
+  test('handles view within innacessible view (iOS)', () => {
```
```suggestion
  test('handles view within inaccessible view (iOS)', () => {
```

mdjastrzebski commented on src/__tests__/to-be-visible.tsx:
```diff
@@ -0,0 +1,127 @@
+    expect(getByTestId('test')).not.toBeVisible();
+  });
+
+  test('handles innacessible view (iOS)', () => {
```
```suggestion
  test('handles inaccessible view (iOS)', () => {
```

fluiddot replied:
Ah, my bad, thanks for flagging the typo 🙇 .

mdjastrzebski commented:
@fluiddot looks pretty solid. I think the code is ready, I've added some tweak to code comments. Pls fix code coverage back to 100%. It should be a matter of adding a single test probably.

fluiddot commented:
> @fluiddot looks pretty solid. I think the code is ready, I've added some tweak to code comments. Pls fix code coverage back to 100%. It should be a matter of adding a single test probably.

Thanks for the review @mdjastrzebski 🙇. I've just applied your suggestions and typo fixes. I'll take a look at the code coverage and try to add another test to provide 100% 👍.

fluiddot commented:
> > @fluiddot looks pretty solid. I think the code is ready, I've added some tweak to code comments. Pls fix code coverage back to 100%. It should be a matter of adding a single test probably.
> 
> Thanks for the review @mdjastrzebski 🙇. I've just applied your suggestions and typo fixes. I'll take a look at the code coverage and try to add another test to provide 100% 👍.

I added a new test for checking that the matcher throws an error when the expectation is not matched. With this, the coverage should go back to 100% 🎊.

mdjastrzebski reviewed (approved):
Looks good! Happy to finally merge it :-) Thank you @fluiddot for this contribution!

fluiddot commented:
> Looks good! Happy to finally merge it :-) Thank you @fluiddot for this contribution!

Thank you so much @mdjastrzebski for the review and help of this contribution 🙇! Happy to know that it's finally merged 😃!
